### PR TITLE
lib!: change CloudEvent to use direct object notation and get/set properties

### PR DESCRIPTION
**Pull request by lance**
This commit makes a substantial change to the API, changing the CloudEvent class
to accept properties as an object in the constructor. For example:

```js
const CloudEvent = require('cloudevents-sdk');

// all event properties except extensions may be set in the constructor
const event = new CloudEvent({
  source: 'http://my.event.source',
  type: 'test-event-type'
});

// get and set all properties standard property notation
console.log(event.time); // the event timestamp
event.subject = 'my event subject';
```

This change includes a new test file to explicitly test this interface,
as well as new documentation.

Fixes: https://github.com/cloudevents/sdk-javascript/issues/65
Fixes: https://github.com/cloudevents/sdk-javascript/issues/148
Fixes: https://github.com/cloudevents/sdk-javascript/issues/135
Fixes: https://github.com/cloudevents/sdk-javascript/issues/88

### CODE REVIEW

**Comment by lance** [comment]
I realize there are a lot of changes in the HTML, however, the HTML is auto-generated, so usually I don't think it's necessary to spend time reviewing it. It's just reflecting what's in the code. Typically, I will just click the "Viewed" checkbox on the PR so they don't get in the way. Since our docs are configured to be published at https://cloudevents.github.io/sdk-javascript/ we need to have them on `master` instead of another branch. :man_shrugging:

**Comment by grant** [comment]
I would really like no autogenerated files in source control. We can think of some other way to get the docs besides this. Not sure if others agree.

**Comment by lance** [comment]
I think it's pretty common to use the `docs/` dir as a place for a project's website. That's what [sdk-go does](https://cloudevents.github.io/sdk-go/), for example. Of course, they have godoc.org to publish their API which we don't have - and which is why our generated documentation lives in `docs/`. I don't think these will be changing regularly - only when there are big API changes like this.

**Comment by lance** [comment]
Since it's such a big change, if you just want to see how the API is used, I added some tests that just exercise the new interface. https://github.com/cloudevents/sdk-javascript/pull/172/files#diff-dd013ce6952212675e25f44a069d648a

File: lib/bindings/http/validation/structured.js [hunk 1]
```diff
@@ -14,8 +14,7 @@ function check(payload, headers, receiver) {
   const sanityHeaders = sanityAndClone(headers);
 
   // Validation Level 1
-  if (!receiver.allowedContentTypes
-    .includes(sanityHeaders[HEADER_CONTENT_TYPE])) {
+  if (!receiver.allowedContentTypes.includes(sanityHeaders[HEADER_CONTENT_TYPE])) {
```
Comment (helio-frota) [inline lib/bindings/http/validation/structured.js:17]: Not a blocker, but I remember something like that in other file (?) : ) 

if ( not_receiver_allowed_includes foo )

File: lib/cloudevent.js [hunk 2]
```diff
@@ -1,232 +1,314 @@
...
+const Spec03 = require("./bindings/http/v03/spec_0_3.js");
 const Formatter = require("./formats/json/formatter.js");
 
+const { SPEC_V1, SPEC_V03 } = require("./bindings/http/constants.js");
+const { isBinary } = require("./bindings/http/validation/fun.js");
+
 /**
- * An instance of a CloudEvent.
+ * An CloudEvent describes event data in common formats to provide
+ * interopability across services, platforms and systems.
```
Comment (helio-frota) [inline lib/cloudevent.js:10]: typo `interopability`

**Comment by helio-frota** [comment]
regarding some core business changes that may exist here and there I have no much idea, other than that it looks good the move from prototype to ES6 classes. LGTM after the minor changes :+1:

File: lib/bindings/http/validation/structured.js (hunk 1 above)
Comment (helio-frota) [reply lib/bindings/http/validation/structured.js:17]: Sorry missing the context here, I remember that we saw something like that in other file.
and you extracted some variables to improve the readability.

Comment (helio-frota) [reply lib/bindings/http/validation/structured.js:17]: @lance I found it https://github.com/cloudevents/sdk-javascript/pull/134#discussion_r422346076

Comment (lance) [reply lib/bindings/http/validation/structured.js:17]: Ahh gotcha. The difference here is that this is for a structured receiver. The PR you point to is modifying a binary event receiver where an event is sent without any content-type header. For structured receivers, [the content-type must be there](https://github.com/cloudevents/spec/blob/v1.0/http-protocol-binding.md#321-http-content-type).

**Pull request merged**